node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,796 @@
+	w.connectionIdLock.Lock()
+	defer w.connectionIdLock.Unlock()
+	connectionID, exists := w.connectionIdMap[channelID]
+	if exists {
+		return connectionID, nil
+	}
+
+	connectionID, err := w.queryConnectionID(channelID)
+	if err != nil {
+		return connectionID, err
+	}
+
+	w.connectionIdMap[channelID] = connectionID
+	return connectionID, nil
+}
+
+// ibcChannelQueryResults is used to parse the result from the IBC connection ID query.
+type ibcChannelQueryResults struct {
+	Channel struct {
+		State          string
+		ConnectionHops []string `json:"connection_hops"`
+		Version        string
+	}
+}
+
+/*
+This query:
+  http://localhost:1319/ibc/core/channel/v1/channels/channel-0/ports/wasm.wormhole1nc5tatafv6eyq7llkr2gv50ff9e22mnf70qgjlv737ktmt4eswrq0kdhcj
+
+Returns something like this:
+{
+  "channel": {
+    "state": "STATE_OPEN",
+    "ordering": "ORDER_UNORDERED",
+    "counterparty": {
+      "port_id": "wasm.terra14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9ssrc8au",
+      "channel_id": "channel-0"
+    },
+    "connection_hops": [
+      "connection-0"
+    ],
+    "version": "ibc-wormhole-v1"
+  },
+  "proof": null,
+  "proof_height": {
+    "revision_number": "0",
+    "revision_height": "358"
+  }
+}
+*/
+
+// getConnectionID queries the contract on wormchain to map a channel ID to a connection ID.
+func (w *Watcher) queryConnectionID(channelID string) (string, error) {
+	// TODO: cache the channelID -> connectionID mapping. This should be safe because once a channel is closed, that number can never be reused.
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+
+	query := fmt.Sprintf("%s/ibc/core/channel/v1/channels/%s/ports/wasm.%s", w.lcdUrl, channelID, w.contractAddress)
+	connResp, err := client.Get(query)
+	if err != nil {
+		w.logger.Error("channel query failed", zap.String("query", query), zap.Error(err))
+		return "", fmt.Errorf("query failed: %w", err)
+	}
+	connBody, err := io.ReadAll(connResp.Body)
+	if err != nil {
+		return "", fmt.Errorf("read failed: %w", err)
+	}
+	connResp.Body.Close()
+
+	var result ibcChannelQueryResults
+	err = json.Unmarshal(connBody, &result)
+	if err != nil {
+		return "", fmt.Errorf("failed to unmarshal response: %s, error: %w", string(connBody), err)
+	}
+
+	if len(result.Channel.ConnectionHops) != 1 {
+		return "", fmt.Errorf("response contained %d hops when it should return exactly one, json: %s", len(result.Channel.ConnectionHops), string(connBody))
+	}
+
+	w.logger.Info("queried connection ID", zap.String("channelID", channelID), zap.String("connectionID", result.Channel.ConnectionHops[0]))
+	return result.Channel.ConnectionHops[0], nil
+}
+
+// parseEvent parses a wasm event. If it is from the desired contract and for the desired action, it returns an event. Otherwise, it returns nil.
+func parseEvent[T any](logger *zap.Logger, desiredContract string, desiredAction string, event gjson.Result) (*T, error) {
```

Comment:
any reason you're not merging this with the below function and avoiding the double serialization and deserialization?

Reply:
Reworked the parsing code.